### PR TITLE
[ORC] Support Filter and Aggregation pushdown for Decimal column type

### PR DESCRIPTION
**Pull request by chenxu14**
**PR Summary**
1. Add SIMD support for **DecimalUtil::fillDecimals**
2. Filter pushdown
processing **IsNull** and **IsNotNull** during decoding phase, other Filters after decoding
3. Aggregation pushdown
currently only support **Min** and **Max**
**DecimalSumAggregate** don't support function pushdown yet

### CODE REVIEW

**Comment by chenxu14** [comment]
Moved UT to OrcTableScanTest, since DWRF don’t support Decimal Type

**Comment by chenxu14** [comment]
linux-presto-fuzzer-run failed due to #6428

File: velox/type/DecimalUtil.h [hunk 1]
```diff
@@ -91,28 +116,126 @@ class DecimalUtil {
...
-      int32_t numValues,
-      int32_t targetScale) {
-    for (int32_t i = 0; i < numValues; i++) {
-      if (!nullsPtr || !bits::isBitNull(nullsPtr, i)) {
-        int32_t currentScale = scales[i];
-        T value = values[i];
-        if constexpr (std::is_same_v<T, std::int64_t>) { // Short Decimal
+      int32_t targetScale,
+      raw_vector<int32_t>& rows,
+      const velox::common::Filter& filter) {
```
Comment (Yuhta) [inline velox/type/DecimalUtil.h:125]: Template the `Filter` type and make sure the call to `testValues` are inlined in column reader.

File: velox/type/DecimalUtil.h [hunk 2]
```diff
@@ -91,28 +116,126 @@ class DecimalUtil {
...
+          }
+          int32_t fstMask = simd::toBitMask(filter.testValues(fstValues));
+          int32_t secMask = simd::toBitMask(filter.testValues(secValues));
+          int32_t mask = fstMask | secMask << halfBatch;
+
+          // save values and outputRows
+          if (mask) {
+            if (mask == simd::allSetBitMask<int32_t>()) {
+              xsimd::load_aligned(&rows[i]).store_aligned(&rows[rowsCnt]);
+              fstValues.store_aligned(decimals + rowsCnt);
```
Comment (Yuhta) [inline velox/type/DecimalUtil.h:171]: Reduce data dependency:
```
fstValues.store_aligned(decimals + rowsCnt);
secValues.store_aligned(decimals + rowsCnt + halfBatch);
```

File: velox/type/DecimalUtil.h [hunk 3]
```diff
@@ -91,28 +116,126 @@ class DecimalUtil {
...
+                  rowsCnt += __builtin_popcount(secMask);
+                }
+              }
+            }
+          }
+        }
+        startIndex = vecSize;
+      }
+      // remaining part can't be vectorize
+      for (int32_t rowIndex = startIndex; rowIndex < numRows; rowIndex++) {
```
Comment (Yuhta) [inline velox/type/DecimalUtil.h:210]: Create an helper function to reuse for both remaining int64 and int128 loops.

File: velox/type/DecimalUtil.h [hunk 4]
```diff
@@ -91,28 +116,126 @@ class DecimalUtil {
...
+            } else {
+              if (isDense) {
+                auto fstRow = rows[rowsCnt];
+                (xsimd::load_unaligned(simd::byteSetBits(mask)) + fstRow)
+                    .store_aligned(&rows[rowsCnt]);
+              } else {
+                simd::filter(xsimd::load_aligned(&rows[i]), mask)
+                    .store_aligned(&rows[rowsCnt]);
+              }
+              if (fstMask) {
```
Comment (Yuhta) [inline velox/type/DecimalUtil.h:184]: Extract helper function for storing values.  Then you can break the data dependency a bit:
```
storeValues(fstMask, decimals + rowsCnt);
storeValues(secMask, decimals + rowsCnt + __builtin_popcount(fstMask));
```

File: velox/dwio/dwrf/reader/SelectiveDecimalColumnReader.cpp [hunk 5]
```diff
@@ -87,56 +84,153 @@ void SelectiveDecimalColumnReader<DataT>::readHelper(RowSet rows) {
...
+      scanSpec_->filter() ? scanSpec_->filter() : &alwaysTrue();
+  if (filter->testNull() && version_ == velox::dwrf::RleVersion_2) {
+    /// RLEv2 does't support fastpath yet, temporarily disable bulkPathEnable_
+    /// to ensure that resultNulls_ will be allocated during prepareNulls()
+    /// or addNull() will be failed if the target Filter accept NULLs
+    bulkPathEnable_ = false;
+  }
   prepareRead<int64_t>(offset, rows, incomingNulls);
+  // enable fastpath after prepare has been finished
+  bulkPathEnable_ = true;
```
Comment (Yuhta) [inline velox/dwio/dwrf/reader/SelectiveDecimalColumnReader.cpp:146]: Toggling it is dangerous.  Just define
```
bool hasBulkPath() const override {
  return version_ != velox::dwrf::RleVersion_2;
}
```
and remove it after fastpath is added in RLEv2.

File: velox/dwio/dwrf/reader/SelectiveDecimalColumnReader.cpp [hunk 6]
```diff
@@ -87,56 +84,153 @@ void SelectiveDecimalColumnReader<DataT>::readHelper(RowSet rows) {
...
   bool isDense = rows.back() == rows.size() - 1;
-  if (isDense) {
-    readHelper<true>(rows);
+  if (scanSpec_->keepValues()) {
+    if (scanSpec_->valueHook()) {
```
Comment (Yuhta) [inline velox/dwio/dwrf/reader/SelectiveDecimalColumnReader.cpp:149]: There is no aggregate pushdown defined for decimals.  Just `VELOX_CHECK(!scanSpec_->valueHook())` for now

**Review by Yuhta — COMMENTED** [review]
Can you add some tests in https://github.com/facebookincubator/velox/blob/main/velox/dwio/dwrf/test/E2EFilterTest.cpp for decimals (both short and long) to ensure the correctness?

**Comment by chenxu14** [comment]
thanks for your review @Yuhta , after talk with [kecookier](https://github.com/kecookier) offline, he will help to inprove this. cc @zhaokuo

**Pull request closed without merging**